MemoryFile

sgillies opened the pull request:
Resolves #899

sgillies commented:
What's this PR about? I've woken up to some use cases for formatted, in-memory, data files. Because reading a megabyte from memory is much faster than reading from disk (see https://gist.github.com/sgillies/f8173c3f6607ba0fb8f02d9964b61cff), when a GeoTIFF is small enough to fit in a computer's RAM we should be able to benefit from storing it in GDAL's in-memory filesystem and accessing it from there with Rasterio.

We can sort of do this already by using magic "/vsimem/" paths:

```python
with rasterio.open('/vsimem/example.tif', 'w', driver='GTiff', ...) as dataset:
    # write data
```

But having to know the unique filename for something that is essentially a temporary file is sort of an anti-feature from Rasterio's perspective. I've come up with an API that is inspired by `tempfile.NamedTemporaryFile()` and `py.path` (used in `py.test`).

Given `data` bytes read from a network socket (not necessary a local file), the temporary disk file usage is this:

```python
with tempfile.NamedTemporaryFile() as tmpfile:
    tmpfile.write(data)
    with rasterio.open(tmpfile.name) as src:
        data = src.read()
```

The `MemoryFile` version is

```python
with MemoryFile(data) as memfile:
    with memfile.open() as src:
        data = src.read()
```

Benchmarks of Rasterio's `MemoryFile` against `tempfile.NamedTemporaryFile` are here: https://gist.github.com/sgillies/af2479c6873aa06b9ecc74643678c53a. Writing the bytes to a `MemoryFile` and opening a dataset from the in-memory file is twice as fast on my MacBook as doing so with a file on its hard disk.

What about writing new datasets? `MemoryFile` handles this, too. You can create an empty `MemoryFile`, create a new formatted, e.g. GeoTIFF, file in it, write arrays of data to the GeoTIFF, and then send its bytes elsewhere without touching a hard disk.

```python
with MemoryFile() as memfile:
    with memfile.open(driver='GTiff', count=3, ...) as dataset:
        dataset.write(arr)
    boto3.client('s3').put_object(Bucket='example', Key='example.tif', Body=memfile)
```

@perrygeo @brendan-ward @geowurster I'd love some review of the idea and code when you have time.

geowurster commented:
@sgillies This looks pretty slick.  I'll check it out in the next few days.

sgillies commented on rasterio/io.py:
```diff
@@ -195,6 +202,66 @@ def __repr__(self):
+    def open(self, driver=None, width=None, height=None,
+             count=None, crs=None, transform=None, dtype=None, nodata=None,
+             **kwargs):
+        """Open the file and return a Rasterio dataset object."""
```
Well, `rasterio.open(memfile.name)` works right now, but it's more verbose than `memfile.open()` and 1) lacks the proper emphasis on the `MemoryFile` as the context, and 2) surfaces the "/vsimem" name unnecessarily. I'm also using http://py.readthedocs.io/en/latest/path.html#py._path.local.LocalPath.open as a guiding light. Its `open()` method returns a Python file object and I find this to be a good analogy for Rasterio.

perrygeo replied:
Meh. I'm -0 on that py.path interface. `memfile.open()` and `path.open()` both feel like thin abstractions over sending a path argument to a familiar method. But we can always fall back on the `.name` property if we need to interoperate with functions that expect a path.

geowurster replied:
The `py.path` style abstraction feels a little foreign to me too.  What if `MemoryFile()` didn't have an `open()` method and instead just let the bytes be optional?  That feels more like `StringIO()` and `BytesIO()`, which I think is a better analogy than file-like object referencing data on disk.

```python
with MemoryFile(bytes) as src, MemoryFile(**src.profile) as dst:
    dst.write(src.read())
    boto3.client('s3').put_object(Bucket='example', Key='example.tif', Body=dst)
```

sgillies commented on rasterio/io.py:
```diff
@@ -195,6 +202,66 @@ def __repr__(self):
+            s.start()
+        except:
+            import pdb; pdb.set_trace()
+            raise
```
👍

sgillies commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
 
+cdef class MemoryFileBase(object):
+    """Base for a BytesIO-like class backed by an in-memory file."""
+
```
I've had the same thought, but there's a big difference between `MemoryFile` and S3 or HTTPs resources. The latter exist outside our Python processes. We can make them or modify them using APIs, yes, but external servers decide their state. A `MemoryFile` exists inside our Python process and not only do we decide its state, we need to deliberately keep it open, because closing a `MemoryFile` unlinks the in-memory file. We need a little extra abstraction in this case, and `MemoryFile` is it.

perrygeo replied:
What about unlocking vsis3/vsihttp writing?

perrygeo replied:
Currently we [can't open vsi data sources in write mode](https://github.com/mapbox/rasterio/blob/master/rasterio/_io.pyx#L888-L890). Now that we have the plumbing to write `vsimem` using the VSI IO functions, could we use apply this same technique to write to `vsis3`?

perrygeo commented on tests/test_memoryfile.py:
```diff
@@ -0,0 +1,114 @@
+
+    with MemoryFile() as memfile:
+        with memfile.open(**profile) as dst:
+            dst.write(data)
```
Y, I was thinking the state of the MemoryFile instance from multiple procs/threads could somehow interfere with the ability to read from it. But looking closer at the implementation, its unlikely (like overlapping-uuid4 unlikely) that would ever happen.

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+            vsi_handle = VSIFileFromMemBuffer(path, buffer,
+                                              len(self._initial_bytes), 0)
+            if vsi_handle == NULL:
+                raise OSError('failed to map buffer to file')
```
@sgillies these errors are going to seem cryptic to users.  How about:
'failed to create in-memory file using initial bytes'

and

'failed to properly close in-memory file'

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+        cdef const char *path = path_b
+
+        cdef VSILFILE *fp = VSIFOpenL(path, 'r')
+        if fp == NULL:
```
you could simplify logic here a bit:

```
if fp != NULL:
    VSIFCloseL(fp)
    return True

return False
```

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+        return self.getbuffer().size
+
+    def close(self):
+        """Close MemoryFile and unlink its in-memory file."""
```
'unlink' doesn't seem immediately obvious, but perhaps makes sense in file system context.  Is the effect that the memory is released (eventually)?  Maybe 'Close MemoryFile and release memory' would be more intuitive.

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+        self.name = os.path.join('/vsimem', str(uuid.uuid4()))
+        self._pos = 0
+
+        self._initial_bytes = initial_bytes
```
I don't follow why you are storing `initial_bytes` in the instance.

sgillies replied:
To prevent the initial bytes from being garbage collected in the following situation:

```python
data = bytes('abc')
memfile = MemoryFile(data)
data = None
```

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+
+        fp = VSIFOpenL(path, 'r')
+        if fp == NULL:
+            raise ValueError("NULL file")
```
Definitely non-intuitive error message.

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+    def check(self):
+        """True if the in-memory file exists"""
+        path_b = self.name.encode('utf-8')
+        cdef const char *path = path_b
```
you are doing a lot of encoding and converting to `char *` - is it possible to save this on the instance at creation time instead?

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+                "Failed to seek to offset %s in %s.", self._pos, self.name)
+
+        cdef unsigned char *buffer = <unsigned char *>CPLMalloc(size)
+        objects_read = VSIFReadL(buffer, 1, size, fp)
```
Should this be in a `try` block to make sure that line 910 is always called even if 906-907 fail?

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+        self._pos += len(result)
+        return result
+
+    def seek(self, offset, whence=0):
```
`from_what` is used in [file seek](https://docs.python.org/3/tutorial/inputoutput.html); I haven't seen `whence`.  Maybe note in docstring that numbers for it match file seek.

sgillies replied:
I got whence from https://docs.python.org/3/library/io.html#io.IOBase.seek

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+            pos = self._pos + offset
+        elif whence == 2:
+            pos = len(self) - offset
+        self._pos = pos
```
Should this prevent being < 0 or >= len(self)?

brendan-ward commented on rasterio/io.py:
```diff
@@ -195,6 +202,66 @@ def __repr__(self):
+
+    A MemoryFile created with initial bytes becomes immutable. A
+    MemoryFile created without initial bytes may be written to using
+    either file-like or dataset interfaces.
```
I'm not seeing where immutability is enforced in the Cython code.  Did I miss something obvious or is this handled by GDAL?

geowurster replied:
@brendan-ward I think that's [handled by GDAL](https://github.com/mapbox/rasterio/pull/915/files#diff-49692333e8667a6b046cbdcad3c1164cR846), although I'm not familiar with GDAL's VSI internals.

sgillies replied:
Yes, handled by GDAL. We do not give GDAL ownership of the the initial bytes, to avoid Python crashers, and so it can't modify them.

geowurster commented on rasterio/_io.pyx:
```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
+            if VSIFCloseL(vsi_handle) != 0:
+                raise OSError('failed to close mapped file handle')
+
+    def check(self):
```
What about just `exists()`?

sgillies commented:
Thanks, all. Great suggestions. I believed I've addressed the review issues.

@perrygeo there's definitely a distinction to be kept between `MemoryFile` and `InMemoryRaster`. I see the latter as being for we Rasterio developers and for use with GDAL API functions that require a GDALDatasetH (like warping, rasterization, and polygonization do). The former is for Rasterio library users who are dealing with formatted files.
